fix(wallet)_: fix circuit breaker to use health checks for fallback

IvanBelyakoff opened the pull request:
Add tests, fix tests failing under stress due to name clashes
Use CircuitBreaker for blockchain RPC calls.

Changes in CircuitBreaker:
- was no circuit for fallback (worked as dummy fallback with no health checks)
- now a separate circuit for fallback

Changed circuits in Collectibles manager:
- was circuit per chain id
- now circuit per (provider + chain id).


Closes #5388

IvanBelyakoff committed:
fix(wallet)_: fix circuit breaker to use health checks for fallback
functions as well.

Add tests, fix tests failing under stress due to name clashes

status-im-auto commented:
### Jenkins Builds
<details>
<summary>Click to see older builds (32)</summary>

| :grey_question: | Commit | :hash: | Finished (UTC) | Duration | Platform | Result |
|-|-|-|-|-|-|-|
| :heavy_check_mark: | 1d7df0dc | [#1](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-5457/1/) | 2024-07-01 16:05:10 | ~4 min | `ios` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-ios-240701-160026-1d7df0-pr5457.zip) |
| :heavy_check_mark: | 1d7df0dc | [#1](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-5457/1/) | 2024-07-01 16:08:00 | ~7 min | `linux` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-desktop-240701-160029-1d7df0-pr5457.zip) |
| :heavy_check_mark: | 1d7df0dc | [#1](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-5457/1/) | 2024-07-01 16:09:22 | ~8 min | `android` | [:package:`aar`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-android-240701-160024-1d7df0-pr5457.aar) |
| :heavy_multiplication_x: | 1d7df0dc | [#1](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-5457/1/) | 2024-07-01 16:13:47 | ~13 min | `tests` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-5457/1//consoleText) |
| | | | | | | |
| :heavy_check_mark: | ca1ab9f5 | [#2](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-5457/2/) | 2024-07-02 18:43:52 | ~2 min | `linux` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-desktop-240702-184123-ca1ab9-pr5457.zip) |
| :heavy_check_mark: | ca1ab9f5 | [#2](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-5457/2/) | 2024-07-02 18:44:07 | ~2 min | `ios` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-ios-240702-184123-ca1ab9-pr5457.zip) |
| :heavy_check_mark: | ca1ab9f5 | [#2](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-5457/2/) | 2024-07-02 18:46:41 | ~5 min | `android` | [:package:`aar`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-android-240702-184118-ca1ab9-pr5457.aar) |
| :heavy_multiplication_x: | ca1ab9f5 | [#2](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-5457/2/) | 2024-07-02 18:48:58 | ~7 min | `tests` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-5457/2//consoleText) |
| | | | | | | |
| :heavy_check_mark: | 6c21146e | [#3](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-5457/3/) | 2024-07-02 19:11:41 | ~1 min | `android` | [:package:`aar`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-android-240702-190949-6c2114-pr5457.aar) |
| :heavy_check_mark: | 6c21146e | [#3](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-5457/3/) | 2024-07-02 19:12:33 | ~2 min | `ios` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-ios-240702-190949-6c2114-pr5457.zip) |
| :heavy_check_mark: | 6c21146e | [#3](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-5457/3/) | 2024-07-02 19:13:42 | ~3 min | `linux` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-desktop-240702-190952-6c2114-pr5457.zip) |
| :heavy_check_mark: | 6c21146e | [#3](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-5457/3/) | 2024-07-02 19:52:48 | ~42 min | `tests` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-5457/3//consoleText) |
| | | | | | | |
| :heavy_check_mark: | bb102fa1 | [#4](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-5457/4/) | 2024-07-03 10:48:57 | ~2 min | `ios` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-ios-240703-104624-bb102f-pr5457.zip) |
| :heavy_check_mark: | bb102fa1 | [#4](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-5457/4/) | 2024-07-03 10:50:03 | ~3 min | `linux` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-desktop-240703-104624-bb102f-pr5457.zip) |
| :heavy_check_mark: | bb102fa1 | [#4](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-5457/4/) | 2024-07-03 10:52:42 | ~6 min | `android` | [:package:`aar`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-android-240703-104624-bb102f-pr5457.aar) |
| :heavy_multiplication_x: | bb102fa1 | [#4](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-5457/4/) | 2024-07-03 10:54:16 | ~7 min | `tests` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-5457/4//consoleText) |
| | | | | | | |
| :heavy_check_mark: | 5df5ff73 | [#5](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-5457/5/) | 2024-07-03 11:45:33 | ~2 min | `ios` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-ios-240703-114313-5df5ff-pr5457.zip) |
| :heavy_check_mark: | 5df5ff73 | [#5](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-5457/5/) | 2024-07-03 11:46:23 | ~3 min | `android` | [:package:`aar`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-android-240703-114313-5df5ff-pr5457.aar) |
| :heavy_check_mark: | 5df5ff73 | [#5](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-5457/5/) | 2024-07-03 11:47:34 | ~4 min | `linux` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-desktop-240703-114314-5df5ff-pr5457.zip) |
| :heavy_check_mark: | 5df5ff73 | [#5](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-5457/5/) | 2024-07-03 12:30:06 | ~46 min | `tests` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-5457/5//consoleText) |
| | | | | | | |
| :heavy_check_mark: | f1f73db2 | [#6](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-5457/6/) | 2024-07-03 13:13:26 | ~2 min | `linux` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-desktop-240703-131112-f1f73d-pr5457.zip) |
| :heavy_check_mark: | f1f73db2 | [#6](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-5457/6/) | 2024-07-03 13:13:28 | ~2 min | `android` | [:package:`aar`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-android-240703-131112-f1f73d-pr5457.aar) |
| :heavy_check_mark: | f1f73db2 | [#6](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-5457/6/) | 2024-07-03 13:14:11 | ~2 min | `ios` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-ios-240703-131112-f1f73d-pr5457.zip) |
| :heavy_check_mark: | f1f73db2 | [#6](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-5457/6/) | 2024-07-03 13:53:49 | ~42 min | `tests` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-5457/6//consoleText) |
| | | | | | | |
| :heavy_check_mark: | 0f8d0e70 | [#7](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-5457/7/) | 2024-07-04 11:34:14 | ~3 min | `ios` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-ios-240704-113109-0f8d0e-pr5457.zip) |
| :heavy_check_mark: | 0f8d0e70 | [#7](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-5457/7/) | 2024-07-04 11:34:30 | ~3 min | `android` | [:package:`aar`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-android-240704-113109-0f8d0e-pr5457.aar) |
| :heavy_multiplication_x: | 0f8d0e70 | [#7](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-5457/7/) | 2024-07-04 11:38:13 | ~6 min | `tests` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-5457/7//consoleText) |
| :heavy_check_mark: | 0f8d0e70 | [#7](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-5457/7/) | 2024-07-04 11:38:20 | ~7 min | `linux` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-desktop-240704-113109-0f8d0e-pr5457.zip) |
| | | | | | | |
| :heavy_multiplication_x: | b6102507 | [#8](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-5457/8/) | 2024-07-04 12:45:25 | ~1 min | `tests` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-5457/8//consoleText) |
| :heavy_check_mark: | b6102507 | [#8](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-5457/8/) | 2024-07-04 12:46:26 | ~2 min | `android` | [:package:`aar`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-android-240704-124355-b61025-pr5457.aar) |
| :heavy_check_mark: | b6102507 | [#8](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-5457/8/) | 2024-07-04 12:46:40 | ~2 min | `ios` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-ios-240704-124355-b61025-pr5457.zip) |
| :heavy_check_mark: | b6102507 | [#8](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-5457/8/) | 2024-07-04 12:47:50 | ~3 min | `linux` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-desktop-240704-124357-b61025-pr5457.zip) |
</details>

| :grey_question: | Commit | :hash: | Finished (UTC) | Duration | Platform | Result |
|-|-|-|-|-|-|-|
| :heavy_check_mark: | c1b305e2 | [#9](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-5457/9/) | 2024-07-04 12:49:53 | ~1 min | `android` | [:package:`aar`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-android-240704-124818-c1b305-pr5457.aar) |
| :heavy_check_mark: | c1b305e2 | [#9](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-5457/9/) | 2024-07-04 12:50:46 | ~2 min | `linux` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-desktop-240704-124822-c1b305-pr5457.zip) |
| :heavy_check_mark: | c1b305e2 | [#9](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-5457/9/) | 2024-07-04 12:51:33 | ~3 min | `ios` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-ios-240704-124818-c1b305-pr5457.zip) |
| :heavy_check_mark: | c1b305e2 | [#9](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-5457/9/) | 2024-07-04 13:32:09 | ~43 min | `tests` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-5457/9//consoleText) |
| | | | | | | |
| :heavy_check_mark: | 66db8115 | [#1](https://ci.status.im/job/status-go/job/prs/job/tests-rpc/job/PR-5457/1/) | 2024-07-04 15:28:32 | ~2 min | `tests-rpc` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/tests-rpc/job/PR-5457/1//consoleText) |
| :heavy_check_mark: | 66db8115 | [#10](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-5457/10/) | 2024-07-04 15:29:02 | ~3 min | `android` | [:package:`aar`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-android-240704-152544-66db81-pr5457.aar) |
| :heavy_check_mark: | 66db8115 | [#10](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-5457/10/) | 2024-07-04 15:29:16 | ~3 min | `ios` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-ios-240704-152544-66db81-pr5457.zip) |
| :heavy_check_mark: | 66db8115 | [#10](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-5457/10/) | 2024-07-04 15:29:33 | ~3 min | `linux` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-desktop-240704-152544-66db81-pr5457.zip) |
| :heavy_multiplication_x: | 66db8115 | [#10](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-5457/10/) | 2024-07-04 15:35:05 | ~9 min | `tests` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-5457/10//consoleText) |
| :heavy_check_mark: | 66db8115 | [#11](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-5457/11/) | 2024-07-04 16:19:29 | ~43 min | `tests` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-5457/11//consoleText) |

IvanBelyakoff committed:
feat(wallet)_: use CircuitBreaker for blockhain RPC calls

fix usage of circuit breaker for collectibles and market data to
match the implementation

IvanBelyakoff commented on rpc/chain/client.go:
```diff
@@ -253,319 +242,257 @@ func (c *ClientWithFallback) makeCall(ctx context.Context, main func() ([]any, e
-					return nil
-				}
+				// Should we handle it differently? Do we allow trying the next provider?
+				// if isVMError(err) {
```
Do we need to handle it differently? Why did not we allow to try next provider?

alaibe replied:
That one is quite import to to keep, we have difficult time to understand the real reason of the error,
in those cases you can retry with any provider, the same error will happens

IvanBelyakoff commented on circuitbreaker/circuit_breaker.go:
```diff
@@ -54,82 +56,81 @@ type CircuitBreaker struct {
+			result.err = fmt.Errorf("%s.error: %w", circuitName, err)
 		}
+		result.res = nil
+		// Lets abuse every provider with the same amount of MaxConcurrentRequests,
```
Are you OK with this?

alaibe replied:
Can u explain more?

IvanBelyakoff replied:
we set MaxConcurrentRequests for CircuitBreaker config. But we have multiple circuits for one command - one circuit for main provider, one for each fallback. This MaxConcurrentRequests is our limitation per circuit. So once we reach it for the main provider we could stop and return or we could forward the rest requests to the next provider.
I will leave it like this for now (works as kind of load balancing). We can change it if that causes any issues

IvanBelyakoff commented on rpc/chain/client.go:
```diff
@@ -129,70 +138,50 @@ var propagateErrors = []error{
 		Timeout:               10000,
 		MaxConcurrentRequests: 100,
-		SleepWindow:           300000,
+		SleepWindow:           30000,
```
5 minutes too long? Should we retry to open circuit in 30 seconds?

IvanBelyakoff commented on rpc/chain/client.go:
```diff
@@ -129,70 +138,50 @@ var propagateErrors = []error{
+func NewClient(ethClients []*EthClient, chainID uint64) *ClientWithFallback {
+	cbConfig := circuitbreaker.Config{
+		CircuitName:           fmt.Sprintf("ethClient_%d", chainID),
+		Timeout:               10000,
```
What calls can take that long? `getLogs` ? calls to contracts to fetch balances?

For such commands we can use a separate circuit with its own config, benefits:
- we get most failures with `getLogs` command. If we have a separate circuit for it, it wont affect the main circuit health for the rest commands
- we can have a lesser timeout for the rest command so if main provider does not respond, we will switch faster to the fallback

WDYT?

And I suggest to remove toggling connection state for provider due to `getLogs` failure, as it normally just harms leading to provider Up/down banner, which is often misleading (the rest of commands - non-archival requests can likely succeed if we have not reached some other limit)

alaibe replied:
Mobile connection?

IvanBelyakoff replied:
But for the rest? disabling connection state switch for getLogs command? A separate circuit for getLogs command? Are you ok with this?

IvanBelyakoff reviewed (commented):
Guys I need your feedback on some points I mentioned inline

alaibe commented on rpc/client.go:
```diff
@@ -109,7 +110,11 @@ func NewClient(client *gethrpc.Client, upstreamChainID uint64, upstream params.U
 			return nil, fmt.Errorf("get RPC limiter: %s", err)
 		}
-		c.upstream = chain.NewSimpleClient(limiter, upstreamClient, upstreamChainID)
+		hostPortUpstream, err := extractHostAndPortFromURL(c.upstreamURL)
```
pretty sure upstream is now fully deprecated

IvanBelyakoff replied:
Ok, will remove it. Do we start removing it everywhere whenever we touch the code step by step?

alaibe replied:
better step by step and confirm with @J-Son89

IvanBelyakoff replied:
Created an issue for this https://github.com/status-im/status-go/issues/5476

saledjenic commented on circuitbreaker/circuit_breaker.go:
```diff
@@ -54,82 +56,81 @@ type CircuitBreaker struct {
 
 func NewCircuitBreaker(config Config) *CircuitBreaker {
+	if config.CircuitName == "" {
+		return nil
```
Could we return `error` in this case (cause there is no place where we check if the created instance is nil or not)?

IvanBelyakoff replied:
I gave it a second thought and I will remove CircuitName from config completely. It should be passed with every Functor only, as this satisfies most cases and is less error prone. Thanx for the heads up!

saledjenic commented on circuitbreaker/circuit_breaker.go:
```diff
@@ -54,82 +56,81 @@ type CircuitBreaker struct {
-				return result
-			}
+		// Accumulate errors
+		if result.err != nil {
```
This won't be `true` ever, look at the condition in the line `110`.

IvanBelyakoff replied:
we have a loop here. when first function fails, we return err from it, which is set to result.err at line 124. Then during next iteration if the function fails, result.err is not nil. I know it works, as it is covered with tests.

saledjenic replied:
What I want to say, you have this code:
```Go
		err := hystrix.DoC(ctx, circuitName, func(ctx context.Context) error {
			res, err := f.exec()
			// Write to result only if success
			if err == nil {
				result = CommandResult{res: res, err: err}
			}
			return err
		}, nil)
```
This means you set `err` of `CommandResult` only when the error is `nil`, which is useless, cause it's nil by default.

Then in the line `124` you have this check `if result.err != nil {` which results in `false`.

saledjenic replied:
Also in the next iteration, you're not dealing with an error from that iteration, but with an error from the previous iteration cause you don't have `result.err = nil` in line `129`, which means if you have many commands you will always have in `result.err` an error from the previous iteration.

saledjenic replied:
To me, something like this, has the most sense:
```Go
		err := hystrix.DoC(ctx, circuitName, func(ctx context.Context) error {
			res, err := f.exec()
			result = CommandResult{res: res}
			return err
		}, nil)

		if err == nil {
			break
		}
		result.err = fmt.Errorf("%s.error: %w", circuitName, err)
```

IvanBelyakoff replied:
if `err == nil` here:
```
			if err == nil {
				result = CommandResult{res: res, err: err}
			}
```
we break at line 110 and never get to line 124.

But if `err != nil` there, then it is returned. if it is the first function in the list, then we get to line 115, if not (means that first failed and already set result.err to err, and the second fails as well, we get to 113.

Again, this is tested in tests, line 70-72 that all errors are accumulatd:
```
	assert.True(t, errors.Is(result.Error(), hystrix.ErrTimeout))
	assert.True(t, errors.Is(result.Error(), errThirdProvFailed))
	assert.True(t, errors.Is(result.Error(), errThirdProvFailed))
```

saledjenic replied:
Again, it works because of this part, but entire implementation is a bit messy
```Go
		if result.err != nil {
			result.err = fmt.Errorf("%w, %s.error: %w", result.err, circuitName, err)
		} else {
			result.err = fmt.Errorf("%s.error: %w", circuitName, err)
		}
```

saledjenic replied:
For example this is simplified:
```Go
func (cb *CircuitBreaker) Execute(cmd Command) CommandResult {
	var result CommandResult
	ctx := cmd.ctx
	if ctx == nil {
		ctx = context.Background()
	}

	for i := 0; i < len(cmd.functors); i += 1 {
		f := cmd.functors[i]

		circuitName := f.circuitName
		if circuitName == "" {
			circuitName = cb.config.CircuitName
		}
		if hystrix.GetCircuitSettings()[circuitName] == nil {
			hystrix.ConfigureCommand(circuitName, hystrix.CommandConfig{
				Timeout:                cb.config.Timeout,
				MaxConcurrentRequests:  cb.config.MaxConcurrentRequests,
				RequestVolumeThreshold: cb.config.RequestVolumeThreshold,
				SleepWindow:            cb.config.SleepWindow,
				ErrorPercentThreshold:  cb.config.ErrorPercentThreshold,
			})
		}

		err := hystrix.DoC(ctx, circuitName, func(ctx context.Context) error {
			res, err := f.exec()
			result.res = res
			return err
		}, nil)

		if err == nil {
			result.err = nil
			break
		}
		result.err = fmt.Errorf("%w, %s.error: %w", result.err, circuitName, err)
	}

	return result
}
```

saledjenic replied:
Even I am not sure if we want this at all `result.err = fmt.Errorf("%w, %s.error: %w", result.err, circuitName, err)`, it's enough to be just `result.err = err`

Cause in that case we could remove this block entirely:
```
		if err == nil {
			result.err = nil
			break
		}
		result.err = fmt.Errorf("%w, %s.error: %w", result.err, circuitName, err)
```

and have only:
```
result.err = hystrix.DoC(....)
if result.err == nil {
  break
}
```

IvanBelyakoff replied:
Sale, you can run the tests with your implementation and they will fail. Because of one reason - it does not accumulate errors, as I said before. you return only one error. For us it is very important to know why each provider failed when the command failed. This is important to resolve bugs where we face RPC errors from providers. Previously it worked fine, because hystrix accumulated the errors, now since the implementation changed, we dont rely on hystrix "fallback" to accumulate errors, we do it ourselves.

saledjenic replied:
That's not true.
Try to use this https://github.com/status-im/status-go/pull/5457#discussion_r1663925537 and you will see.

saledjenic replied:
The changes I posted above accumulate errors.
Which test is failing?

IvanBelyakoff replied:
> That's not true. Try to use this [#5457 (comment)](https://github.com/status-im/status-go/pull/5457#discussion_r1663925537) and you will see.

I did, copied, pasted, run, it failed. Now I see you fixed it (because I have the old version in my code). Now it passes.
But the whole story is about instead of:
```
...
if err == nil {
  result = CommandResult{res: res, err: err}
}
...
if err == nil {
    break
}
if result.err != nil {
 	result.err = fmt.Errorf("%w, %s.error: %w", result.err, f.circuitName, err)
} else {
 	result.err = fmt.Errorf("%s.error: %w", f.circuitName, err)
}

```
We have:
```
...
result.res = res
...
if err == nil {
	result.err = nil
	break
}

result.err = fmt.Errorf("%w, %s.error: %w", result.err, f.circuitName, err)
```

I dont think it was worth the time you  and me wasted

IvanBelyakoff replied:
And btw, if we use only one line, it give the following log in some cases, which just does not look good.
circuit_breaker_test.go:218: %!w(<nil>), CommandCancel_9314320001.error: provider 1 failed

saledjenic commented on circuitbreaker/circuit_breaker.go:
```diff
@@ -54,82 +56,81 @@ type CircuitBreaker struct {
+		} else {
+			result.err = fmt.Errorf("%s.error: %w", circuitName, err)
 		}
+		result.res = nil
```
Don't we need to add this line as well `result.err = nil` otherwise we're referring to the error from the previous functor not the last one?

IvanBelyakoff replied:
Not sure what you mean, In case for any of the functors err is not nil, result.res is set there and we break before we reach here. So, I think this line is just useless here actually. Thanx, will remove it.

saledjenic replied:
It depends how you implement body in `err := hystrix.DoC(ctx, circuitName, func(ctx context.Context) error {.....`

saledjenic commented on services/wallet/collectibles/manager.go:
```diff
@@ -1163,3 +1157,9 @@ func createStatusNotifier(statuses *sync.Map, feed *event.Feed) *connection.Stat
+// Different providers have API keys per chain or per testnet/mainnet.
+// Proper implementation should respect that. For now, the safest solution is to use the provider ID and chain ID as the key.
+func getCircuitName(provider thirdparty.CollectibleProvider, chainID walletCommon.ChainID) string {
+	return provider.ID() + chainID.String()
```
Should we add any delimiter in between, maybe better?
`return fmt.Sprintf("%s-%s", return provider.ID(), chainID.String())`

IvanBelyakoff replied:
I can do that. It does not matter much as this is not for human eyes, just a proper ID

saledjenic replied:
I agree, just thought about it, if I need to print it somewhere, easier for eyes, not a big deal.

IvanBelyakoff replied:
I gave it a second thought and since string concatenations normally a way faster than Sprintf and we use it quite often, and given that this is not for human eyes, I think current implementation is OK

IvanBelyakoff committed:
chore(wallet)_: Added a test for Market manager FetchTokenMarketValues

Added a nil/empty command test for circuit breaker

IvanBelyakoff commented:
Code coverage issues:
- changes in chain.Client now make use of a well tested circuit breaker, no changes in logic
- chain.Client itself serves like a wrapper around calls to providers, so it just forwards calls to providers and sets state to connected/disconnected. Covering it with tests is pretty useless for us.
- some tests were not counted because no changes in the tested code
- in fact we have a better coverage now than before this PR, so the number is not really correct

Please force-merge, @cammellos

igor-sirotin commented:
I'm ok on force-merging
